cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,180 @@
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("could not parse input")
+	}
+	return opts
+}
+
+func (o *options) validate() error {
+	if o.ciOperatorconfigPath == "" {
+		return fmt.Errorf("--ci-operator-config-path must be set")
+	}
+	return nil
+}
+
+func tagsToPromote(cfg *api.ReleaseBuildConfiguration) sets.String {
+	tags := sets.NewString()
+	for _, tag := range release.PromotedTags(cfg) {
+		tags.Insert(fmt.Sprintf("%s/%s:%s", tag.Namespace, tag.Name, tag.Tag))
+	}
+	return tags
+}
+
+func addSchemes() error {
+	if err := imagev1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add imagev1 to scheme: %w", err)
+	}
+	if err := buildv1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add buildv1 to scheme: %w", err)
+	}
+	return nil
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	opts := newOpts()
+
+	if err := opts.validate(); err != nil {
+		logrus.WithError(err).Fatal("failed to validate the option")
+	}
+
+	if err := addSchemes(); err != nil {
+		logrus.WithError(err).Fatal("failed to add schemes")
+	}
+
+	abs, err := filepath.Abs(opts.ciOperatorconfigPath)
+	if err != nil {
+		logrus.WithError(err).WithField("opts.ciOperatorconfigPath", opts.ciOperatorconfigPath).Fatal("failed to determine absolute filepath")
+	}
+	promotedTags := sets.NewString()
+	if err := config.OperateOnCIOperatorConfigDir(abs, func(cfg *api.ReleaseBuildConfiguration, metadata *config.Info) error {
+		promotedTags.Insert(tagsToPromote(cfg).List()...)
+		return nil
+	}); err != nil {
+		logrus.WithField("path", abs).Fatal("failed to operate on CI Operator's config directory")
+	}
+
+	kubeConfigs, _, err := util.LoadKubeConfigs(opts.kubeconfig, nil)
+	if err != nil {
+		logrus.Fatalf("could not load kube config from path %s: %w", opts.kubeconfig, err)
+	}
+	if len(kubeConfigs) != 1 {
+		logrus.Fatalf("found %d contexts in kube config %s: must be %s only", len(kubeConfigs), opts.kubeconfig, string(api.ClusterAPPCI))
+	}
+
+	client, err := ctrlruntimeclient.New(kubeConfigs[string(api.ClusterAPPCI)], ctrlruntimeclient.Options{})
+	if err != nil {
+		logrus.WithError(err).Fatalf("could not create client")
+	}
+
+	ctx := context.TODO()
+	buildConfigs := &buildv1.BuildConfigList{}
+	if err := client.List(ctx, buildConfigs); err != nil {
+		logrus.WithError(err).Fatalf("could not list buildConfigs")
+	}
+	buildConfigOutputTags := sets.NewString()
+	for _, bc := range buildConfigs.Items {
+		if bc.Spec.Output.To != nil {
+			buildConfigOutputTags.Insert(fmt.Sprintf("%s/%s", bc.Spec.Output.To.Namespace, bc.Spec.Output.To.Name))
+		}
+	}
+	tagsToCheck := sets.NewString()
+	var errs []error
+	imageStreamsWithPromotedTags := sets.NewString()
+	for _, promotedTag := range promotedTags.List() {
+		namespacedImageStreamAndTagName := strings.Split(promotedTag, ":")
```

Comment:
Why do we collect `ImageStreamTagReference` objects, which have fields, only to string-format them into dimensionless strings, only to split them back into the fields we started with? Just use a set of `ImageStreamTagReference`s. Or a list, since we have other validation that ensures no two configurations promote the same tag.

Reply:
Use a list. Also added a function to avoid promotion conflict.